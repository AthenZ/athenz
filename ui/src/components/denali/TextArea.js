/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import { rgba } from 'polished';
import { colors, cssFontSizes, cssFontStyles } from './styles/index';

// NOTE: We probably don't need a wrapper div, but keep one in case
// we have something like a message div
const makeCssTextArea = (props) => css`
    ${cssFontStyles.default};
    box-sizing: border-box;
    display: inline-flex;
    flex-flow: column nowrap;
    position: relative;
    width: ${props.fluid ? '100%' : props.width || '250px'};
    label: textarea;

    & > textarea {
        appearance: none;
        background: ${props.dark
            ? rgba(colors.grey300, 0.1)
            : rgba(colors.brand700, 0.05)};
        border-top: 2px solid transparent;
        border-bottom: 2px solid ${props.error ? colors.red600 : 'transparent'};
        border-right: transparent;
        border-left: transparent;
        border-radius: 2px;
        box-shadow: none;
        box-sizing: border-box;
        color: ${props.dark ? colors.white : colors.grey800};
        flex: 1 0 auto;
        font: inherit;
        height: ${props.height};
        line-height: 1.5;
        margin: 0;
        outline: none;
        padding: 0.25em 0.75rem;
        text-align: left;
        transition: ${!props.noanim
            ? 'background-color 0.2s ease-in-out, color 0.2s ease-in-out, border 0.2s ease-in-out'
            : undefined};
        width: auto;
    }

    /* Placeholder */
    & > textarea::placeholder {
        color: ${props.dark
            ? rgba(colors.white, 0.2)
            : rgba(colors.grey800, 0.6)};
    }

    /* Focused / Active */
    & > textarea:active,
    & > textarea:focus {
        background: ${props.dark
            ? rgba(colors.grey300, 0.1)
            : rgba(colors.brand700, 0.05)};
        border-bottom: 2px solid ${colors.brand700};
        color: ${props.dark ? colors.white : colors.grey800};
    }

    /* Invalid */
    & > textarea:invalid {
        background: ${rgba(colors.brand700, 0.05)};
        border-bottom: 2px solid ${colors.red600};
        color: ${colors.grey800};
    }

    /* Disabled */
    & > textarea[disabled] {
        background: ${props.dark
            ? rgba(colors.white, 0.1)
            : rgba(colors.grey800, 0.05)};
        border-bottom: 2px solid
            ${props.dark
                ? rgba(colors.white, 0.05)
                : rgba(colors.grey800, 0.05)};
        color: ${props.dark
            ? rgba(colors.white, 0.1)
            : rgba(colors.grey800, 0.25)};
        cursor: not-allowed;
    }

    & .message {
        ${cssFontSizes.subtitle};
        color: ${props.error ? colors.red600 : colors.black};
        left: 0;
        line-height: 1.8;
        position: absolute;
        top: 2.571rem;
    }
`;

/**
 * Styled wrapper around `<textarea>`
 *
 * Aside from custom props below, all standard `<textarea>` props are supported.
 * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea
 *
 * There are two key differences between using HTML's `<textarea>` and this
 * `<TextArea>` component:
 *
 * 1. `<TextArea>` uses the `value` prop instead of `children`, which is
 * inline with React practices:
 * https://reactjs.org/docs/forms.html#the-textarea-tag
 * 2. For read-only, use `readOnly` instead of `readonly` (note the camel case)
 */
class TextArea extends React.PureComponent {
    render() {
        // Pull out all the component-specific props. This allows the developer to
        // pass in additional `<textarea>` attributes without specifically having to
        // define them within this component.
        const {
            className,
            dark,
            error,
            fluid,
            height,
            innerRef,
            noanim,
            width,
            ...rest
        } = this.props;

        const cssProps = { dark, error, fluid, height, noanim, width };
        const cssTextArea = makeCssTextArea(cssProps);

        return (
            <div
                className={cx(cssTextArea, 'denali-textarea', className)}
                data-testid='textarea-wrapper'
            >
                <textarea {...rest} ref={innerRef} data-testid='textarea' />
            </div>
        );
    }
}

TextArea.propTypes = {
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** Dark theme */
    dark: PropTypes.bool,
    /** Display rrror state */
    error: PropTypes.bool,
    /** TextArea fills entire width of parent container */
    fluid: PropTypes.bool,
    /** Height of the textarea (alternative to `rows` attribute) */
    height: PropTypes.string,
    /** Forward React ref to the `<textarea>` element */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Content of the textarea. This is the single source of truth */
    value: PropTypes.string,
    /** Width of the textarea (alternative to `cols` attribute) */
    width: PropTypes.string,
};

TextArea.defaultProps = {
    dark: false,
    error: false,
    fluid: false,
    height: 'auto',
    noanim: false,
};

export default TextArea;
