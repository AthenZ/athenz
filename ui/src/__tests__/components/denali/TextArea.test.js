/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import TextArea from '../../../components/denali/TextArea';

describe('TextArea', () => {
    const onChange = jest.fn();

    afterEach(() => onChange.mockClear());

    it('should render', () => {
        const { getByTestId } = render(
            <TextArea
                name='test1'
                placeholder='Insert something here'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');
        const textarea = getByTestId('textarea');

        expect(wrapper).toMatchSnapshot();
        expect(textarea.value).toBe('');
    });

    it('should handle width / height', () => {
        const { getByTestId } = render(
            <TextArea
                height='100px'
                name='test2a'
                value=''
                width='400px'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');
        const textarea = getByTestId('textarea');

        expect(wrapper).toMatchSnapshot();
        expect(wrapper).toHaveStyle(`width: 400px`);
        expect(textarea).toHaveStyle(`height: 100px`);
    });

    it('should handle fluid', () => {
        const { getByTestId } = render(
            <TextArea name='test2b' value='' fluid onChange={onChange} />
        );
        const wrapper = getByTestId('textarea-wrapper');

        expect(wrapper).toHaveStyle(`width: 100%`);
    });

    it('should render a disabled TextArea', () => {
        const { getByTestId } = render(
            <TextArea
                disabled
                name='test4'
                placeholder="Can't touch this"
                value=''
                onChange={onChange}
            />
        );
        const textarea = getByTestId('textarea');

        expect(textarea).toHaveAttribute('disabled');
    });

    it('should render dark theme', () => {
        const { getByTestId } = render(
            <TextArea dark name='test5a' value='' onChange={onChange} />
        );
        const wrapper = getByTestId('textarea-wrapper');

        expect(wrapper).toMatchSnapshot();
    });

    it('should render disabled dark theme', () => {
        const { getByTestId } = render(
            <TextArea
                dark
                disabled
                name='test5b'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');
        const textarea = getByTestId('textarea');

        expect(wrapper).toMatchSnapshot();
        expect(textarea).toHaveAttribute('disabled');
    });

    it('should disable transitions', () => {
        const { getByTestId } = render(
            <TextArea
                name='test6'
                noanim
                placeholder='Insert something here'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');

        expect(wrapper).toMatchSnapshot();
    });

    it('should render error state', () => {
        const { getByTestId } = render(
            <TextArea
                error
                name='test7'
                value='This is an error'
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with additional className', () => {
        const { getByTestId } = render(
            <TextArea
                className='custom-class'
                name='test8'
                placeholder='Insert something here'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('textarea-wrapper');

        expect(wrapper).toHaveClass('denali-textarea');
        expect(wrapper).toHaveClass('custom-class');
    });

    it('should handle change with onChange()', () => {
        const { getByTestId } = render(
            <TextArea name='test9' value='old value' onChange={onChange} />
        );
        const textarea = getByTestId('textarea');

        fireEvent.change(textarea, { target: { value: 'new value' } });

        expect(onChange).toHaveBeenCalled();
    });
});
