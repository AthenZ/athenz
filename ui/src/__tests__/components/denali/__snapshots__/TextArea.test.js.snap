// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea should disable transitions 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 250px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    name="test6"
    placeholder="Insert something here"
  />
</div>
`;

exports[`TextArea should handle width / height 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 400px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: 100px;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    name="test2a"
  />
</div>
`;

exports[`TextArea should render 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 250px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    name="test1"
    placeholder="Insert something here"
  />
</div>
`;

exports[`TextArea should render dark theme 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 250px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(242,242,242,0.1);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #fff;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea::placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(242,242,242,0.1);
  border-bottom: 2px solid #3570f4;
  color: #fff;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(255,255,255,0.1);
  border-bottom: 2px solid rgba(255,255,255,0.05);
  color: rgba(255,255,255,0.1);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    name="test5a"
  />
</div>
`;

exports[`TextArea should render disabled dark theme 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 250px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(242,242,242,0.1);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #fff;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea::placeholder {
  color: rgba(255,255,255,0.2);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(242,242,242,0.1);
  border-bottom: 2px solid #3570f4;
  color: #fff;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(255,255,255,0.1);
  border-bottom: 2px solid rgba(255,255,255,0.05);
  color: rgba(255,255,255,0.1);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    disabled=""
    name="test5b"
  />
</div>
`;

exports[`TextArea should render error state 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 250px;
}

.emotion-0 > textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid #d01111;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-0 > textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-0 > textarea:active,
.emotion-0 > textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-0 > textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-0 > textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-0 .message {
  font-size: 12px;
  color: #d01111;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

<div
  class="emotion-0 denali-textarea"
  data-testid="textarea-wrapper"
>
  <textarea
    data-testid="textarea"
    name="test7"
  >
    This is an error
  </textarea>
</div>
`;
